### PR TITLE
Svg refactoring

### PR DESCRIPTION
**Pull request by Ian-Stewart-Binks**
This pull request refactors the Svg Parsing code. There is still *much more* refactoring to be done, but this is a large step in the right direction.

The parsing phase and the database phase have been separated.

#428 should be merged before this is considered.

### CODE REVIEW

File: hs/SvgParsing/parser.hs [hunk 1]
```diff
@@ -18,140 +18,120 @@ import Text.XML.HaXml.Namespaces
...
+             parentStyle = Style adjustedTransform
+                                 newFill
+                                 newStroke
 
-           parseElements (parseRect parentStyle) rects
-           parseElements (parseText parentStyle) texts
-           parseElements (parsePath (currentlyInRegion || isRegion) parentStyle) paths
-           parseElements (parseEllipse parentStyle) ellipses
-           parseChildren (currentlyInRegion || isRegion) parentStyle children
+         let x1= map (parseRect parentStyle) (tag "rect" content)
```
Comment (Ian-Stewart-Binks) [inline hs/SvgParsing/parser.hs]: This section needs to be cleaned up.

File: hs/SvgParsing/parser.hs [hunk 2]
```diff
@@ -18,140 +18,120 @@ import Text.XML.HaXml.Namespaces
...
-              parseLevel False (Style (0,0) "" "") (getRoot graphDoc)
-              liftIO $ print "Parsing complete"
+              mapM_ insert_ shapes
+              mapM_ insert_ paths
+              mapM_ insert_ texts
+          printDB
+          buildSVG M.empty "../res/graphs/CSC/csc_graph.svg"
           createDirectoryIfMissing True "../res/graphs/CSC"
-          buildSVG
           liftIO $ print "SVG Built"
```
Comment (Ian-Stewart-Binks) [inline hs/SvgParsing/parser.hs]: `liftIO`.

**Comment by david-yz-liu** [comment]
@Ian-Stewart-Binks I don't like the fact that the "concurrent counter" is in this pull request. If you remove that, I'll review the rest and merge it in.

File: hs/Database/Tables.hs [hunk 3]
```diff
@@ -70,44 +66,39 @@ Distribution
...
-Rects
-    gId Int
+Text
+--    gId Int
     rId String
-    width Rational
-    height Rational
-    xPos Rational
-    yPos Rational
+    coord Point
```
Comment (david-yz-liu) [inline hs/Database/Tables.hs]: I like `pos`

File: hs/Database/Tables.hs [hunk 4]
```diff
@@ -70,44 +66,39 @@ Distribution
...
+    width Double
+    height Double
     fill String
     stroke String
+    text [Text]
     isHybrid Bool
-    deriving Show
+    tolerance Double
+    isBool Bool
```
Comment (david-yz-liu) [inline hs/Database/Tables.hs]: Why don't we use a `type` string instead of separate attributes for hybrids and bools?

File: hs/ImageResponse.hs [hunk 5]
```diff
@@ -10,41 +10,55 @@ import Control.Monad.IO.Class  (liftIO)
...
-	liftIO $ createImageFile "Testfile2.svg" "INSERT_ID-graph.png"
-	imageData <- BS.readFile "INSERT_ID-graph.png"
-	liftIO $ removeImage "INSERT_ID-graph.png"
-	liftIO $ removeImage "Testfile2.svg"
-	let encodedData = BEnc.encode imageData
-	return $ toResponse encodedData
+getGraphImage :: MVar Integer -> M.Map String String -> IO Response
+getGraphImage counter courseMap =
+    do c <- addCounter counter
+       let svgFilename = (show c ++ "-graph-svg-file.svg")
```
Comment (david-yz-liu) [inline hs/ImageResponse.hs]: no need to parentheses

File: hs/SvgParsing/Builder.hs [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    where coords = pathPoints entity
+
+-- | Builds a Rect from a database entry in the rects table.
+buildRect :: [Text] -> Shape -> Shape
+buildRect texts entity =
+    let rectTexts = filter (\x -> intersects
+                            (shapeWidth entity)
+                            (shapeHeight entity)
+                            (shapeCoord entity)
+                            9
```
Comment (david-yz-liu) [inline hs/SvgParsing/Builder.hs]: Is this now stored in the database?

**Comment by Ian-Stewart-Binks** [comment]
Will do.

File: hs/SvgParsing/Builder.hs (hunk 6 above)
Comment (Ian-Stewart-Binks) [inline hs/SvgParsing/Builder.hs]: Not yet.

File: hs/Database/Tables.hs (hunk 4 above)
Comment (Ian-Stewart-Binks) [inline hs/Database/Tables.hs]: Sure.

File: hs/Database/Tables.hs (hunk 3 above)
Comment (Ian-Stewart-Binks) [inline hs/Database/Tables.hs]: Sure thing.

File: hs/ImageResponse.hs (hunk 5 above)
Comment (Ian-Stewart-Binks) [inline hs/ImageResponse.hs]: Code has been completely removed.

File: hs/SvgParsing/Builder.hs (hunk 6 above)
Comment (Ian-Stewart-Binks) [inline hs/SvgParsing/Builder.hs]: Ah, that, yes I believe it is. I thought you were mentioning the intersecting parts- I'm going to see what I can do about that.

File: hs/Database/Tables.hs (hunk 3 above)
Comment (Ian-Stewart-Binks) [inline hs/Database/Tables.hs]: Done.

**Pull request closed without merging**